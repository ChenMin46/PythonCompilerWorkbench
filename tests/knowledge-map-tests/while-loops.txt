while (x > 0):
    print x
    x -= 1

while (x > 0):
    if x % 2 == 0:
        print x
    x -= 1

while (x > 0):
    if x % 2 == 0:
        print x
    x -= 1

while (x > 0):
    if x % 2 == 0:
        continue
    print x
    x -= 1

while True:
    if x < 0:
        break
    x -= 1

while True:
    pass
