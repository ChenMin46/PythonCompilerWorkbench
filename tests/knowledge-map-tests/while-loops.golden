while (x > 0):
    print x
    x -= 1

while (x > 0):
    if x % 2 == 0:
        print x
    x -= 1

while (x > 0):
    if x % 2 == 0:
        print x
    x -= 1

while (x > 0):
    if x % 2 == 0:
        continue
    print x
    x -= 1

while True:
    if x < 0:
        break
    x -= 1

while True:
    pass
===
Module
  body: [
    While >> 'while (x > 0'
      test: Compare >> 'x > 0'
        left: Name >> 'x'
          id: 'x'
          ctx: Load
        ops: [
          Gt
        ]
        comparators: [
          Num >> '0'
            n: 0
        ]
      body: [
        Print >> 'print x'
          dest: None
          values: [
            Name >> 'x'
              id: 'x'
              ctx: Load
          ]
          nl: True
        AugAssign >> 'x -= 1'
          target: Name >> 'x'
            id: 'x'
            ctx: Store
          op: Sub
          value: Num >> '1'
            n: 1
      ]
      orelse: [
      ]
    While >> 'while (x > 0'
      test: Compare >> 'x > 0'
        left: Name >> 'x'
          id: 'x'
          ctx: Load
        ops: [
          Gt
        ]
        comparators: [
          Num >> '0'
            n: 0
        ]
      body: [
        If >> 'if x % 2 == 0'
          test: Compare >> 'x % 2 == 0'
            left: BinOp >> 'x % 2'
              left: Name >> 'x'
                id: 'x'
                ctx: Load
              op: Mod
              right: Num >> '2'
                n: 2
            ops: [
              Eq
            ]
            comparators: [
              Num >> '0'
                n: 0
            ]
          body: [
            Print >> 'print x'
              dest: None
              values: [
                Name >> 'x'
                  id: 'x'
                  ctx: Load
              ]
              nl: True
          ]
          orelse: [
          ]
        AugAssign >> 'x -= 1'
          target: Name >> 'x'
            id: 'x'
            ctx: Store
          op: Sub
          value: Num >> '1'
            n: 1
      ]
      orelse: [
      ]
    While >> 'while (x > 0'
      test: Compare >> 'x > 0'
        left: Name >> 'x'
          id: 'x'
          ctx: Load
        ops: [
          Gt
        ]
        comparators: [
          Num >> '0'
            n: 0
        ]
      body: [
        If >> 'if x % 2 == 0'
          test: Compare >> 'x % 2 == 0'
            left: BinOp >> 'x % 2'
              left: Name >> 'x'
                id: 'x'
                ctx: Load
              op: Mod
              right: Num >> '2'
                n: 2
            ops: [
              Eq
            ]
            comparators: [
              Num >> '0'
                n: 0
            ]
          body: [
            Print >> 'print x'
              dest: None
              values: [
                Name >> 'x'
                  id: 'x'
                  ctx: Load
              ]
              nl: True
          ]
          orelse: [
          ]
        AugAssign >> 'x -= 1'
          target: Name >> 'x'
            id: 'x'
            ctx: Store
          op: Sub
          value: Num >> '1'
            n: 1
      ]
      orelse: [
      ]
    While >> 'while (x > 0'
      test: Compare >> 'x > 0'
        left: Name >> 'x'
          id: 'x'
          ctx: Load
        ops: [
          Gt
        ]
        comparators: [
          Num >> '0'
            n: 0
        ]
      body: [
        If >> 'if x % 2 == 0'
          test: Compare >> 'x % 2 == 0'
            left: BinOp >> 'x % 2'
              left: Name >> 'x'
                id: 'x'
                ctx: Load
              op: Mod
              right: Num >> '2'
                n: 2
            ops: [
              Eq
            ]
            comparators: [
              Num >> '0'
                n: 0
            ]
          body: [
            Continue >> 'continue'
          ]
          orelse: [
          ]
        Print >> 'print x'
          dest: None
          values: [
            Name >> 'x'
              id: 'x'
              ctx: Load
          ]
          nl: True
        AugAssign >> 'x -= 1'
          target: Name >> 'x'
            id: 'x'
            ctx: Store
          op: Sub
          value: Num >> '1'
            n: 1
      ]
      orelse: [
      ]
    While >> 'while True'
      test: Name >> 'True'
        id: 'True'
        ctx: Load
      body: [
        If >> 'if x < 0'
          test: Compare >> 'x < 0'
            left: Name >> 'x'
              id: 'x'
              ctx: Load
            ops: [
              Lt
            ]
            comparators: [
              Num >> '0'
                n: 0
            ]
          body: [
            Break >> 'break'
          ]
          orelse: [
          ]
        AugAssign >> 'x -= 1'
          target: Name >> 'x'
            id: 'x'
            ctx: Store
          op: Sub
          value: Num >> '1'
            n: 1
      ]
      orelse: [
      ]
    While >> 'while True'
      test: Name >> 'True'
        id: 'True'
        ctx: Load
      body: [
        Pass >> 'pass'
      ]
      orelse: [
      ]
  ]
--- JSON ---
{
  "contents": [
    {
      "contents": [
        {
          "contents": [
            "while (", 
            {
              "contents": [
                {
                  "contents": [
                    "x"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }, 
                " > ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "While"
        }, 
        "):", 
        "\n", 
        "    ", 
        {
          "contents": [
            "print ", 
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }
          ], 
          "id": "id_0", 
          "name": "Print"
        }, 
        "\n    ", 
        {
          "contents": [
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            " -= ", 
            {
              "contents": [
                "1"
              ], 
              "id": "id_0", 
              "name": "Num"
            }
          ], 
          "id": "id_0", 
          "name": "AugAssign"
        }, 
        "\n\n", 
        {
          "contents": [
            "while (", 
            {
              "contents": [
                {
                  "contents": [
                    "x"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }, 
                " > ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "While"
        }, 
        "):", 
        "\n", 
        "    ", 
        {
          "contents": [
            "if ", 
            {
              "contents": [
                {
                  "contents": [
                    {
                      "contents": [
                        "x"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    " % ", 
                    {
                      "contents": [
                        "2"
                      ], 
                      "id": "id_0", 
                      "name": "Num"
                    }
                  ], 
                  "id": "id_0", 
                  "name": "BinOp"
                }, 
                " == ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "If"
        }, 
        ":", 
        "\n", 
        "        ", 
        {
          "contents": [
            "print ", 
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }
          ], 
          "id": "id_0", 
          "name": "Print"
        }, 
        "\n    ", 
        {
          "contents": [
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            " -= ", 
            {
              "contents": [
                "1"
              ], 
              "id": "id_0", 
              "name": "Num"
            }
          ], 
          "id": "id_0", 
          "name": "AugAssign"
        }, 
        "\n\n", 
        {
          "contents": [
            "while (", 
            {
              "contents": [
                {
                  "contents": [
                    "x"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }, 
                " > ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "While"
        }, 
        "):", 
        "\n", 
        "    ", 
        {
          "contents": [
            "if ", 
            {
              "contents": [
                {
                  "contents": [
                    {
                      "contents": [
                        "x"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    " % ", 
                    {
                      "contents": [
                        "2"
                      ], 
                      "id": "id_0", 
                      "name": "Num"
                    }
                  ], 
                  "id": "id_0", 
                  "name": "BinOp"
                }, 
                " == ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "If"
        }, 
        ":", 
        "\n", 
        "        ", 
        {
          "contents": [
            "print ", 
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }
          ], 
          "id": "id_0", 
          "name": "Print"
        }, 
        "\n    ", 
        {
          "contents": [
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            " -= ", 
            {
              "contents": [
                "1"
              ], 
              "id": "id_0", 
              "name": "Num"
            }
          ], 
          "id": "id_0", 
          "name": "AugAssign"
        }, 
        "\n\n", 
        {
          "contents": [
            "while (", 
            {
              "contents": [
                {
                  "contents": [
                    "x"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }, 
                " > ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "While"
        }, 
        "):", 
        "\n", 
        "    ", 
        {
          "contents": [
            "if ", 
            {
              "contents": [
                {
                  "contents": [
                    {
                      "contents": [
                        "x"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    " % ", 
                    {
                      "contents": [
                        "2"
                      ], 
                      "id": "id_0", 
                      "name": "Num"
                    }
                  ], 
                  "id": "id_0", 
                  "name": "BinOp"
                }, 
                " == ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "If"
        }, 
        ":", 
        "\n", 
        "        ", 
        {
          "contents": [
            "continue"
          ], 
          "id": "id_0", 
          "name": "Continue"
        }, 
        "\n    ", 
        {
          "contents": [
            "print ", 
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }
          ], 
          "id": "id_0", 
          "name": "Print"
        }, 
        "\n    ", 
        {
          "contents": [
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            " -= ", 
            {
              "contents": [
                "1"
              ], 
              "id": "id_0", 
              "name": "Num"
            }
          ], 
          "id": "id_0", 
          "name": "AugAssign"
        }, 
        "\n\n", 
        {
          "contents": [
            "while ", 
            {
              "contents": [
                "True"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "While"
        }, 
        ":", 
        "\n", 
        "    ", 
        {
          "contents": [
            "if ", 
            {
              "contents": [
                {
                  "contents": [
                    "x"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }, 
                " < ", 
                {
                  "contents": [
                    "0"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Compare"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "If"
        }, 
        ":", 
        "\n", 
        "        ", 
        {
          "contents": [
            "break"
          ], 
          "id": "id_0", 
          "name": "Break"
        }, 
        "\n    ", 
        {
          "contents": [
            {
              "contents": [
                "x"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            " -= ", 
            {
              "contents": [
                "1"
              ], 
              "id": "id_0", 
              "name": "Num"
            }
          ], 
          "id": "id_0", 
          "name": "AugAssign"
        }, 
        "\n\n", 
        {
          "contents": [
            "while ", 
            {
              "contents": [
                "True"
              ], 
              "id": "id_0", 
              "name": "Name"
            }, 
            ""
          ], 
          "id": "id_0", 
          "name": "While"
        }, 
        ":", 
        "\n", 
        "    ", 
        {
          "contents": [
            "pass"
          ], 
          "id": "id_0", 
          "name": "Pass"
        }
      ], 
      "id": "id_0", 
      "name": "Module"
    }, 
    "\n"
  ], 
  "id": "id_ROOT", 
  "name": "ROOT_NODE"
}
